servicetalk-grpc-protoc/build.gradle

```diff
@@ -47,42 +50,78 @@ shadowJar {
   classifier = ''
 }
 
+def grpcPluginUberJarName = project.name + "-" + project.version + "-all.jar"
+
+task copyUberJarForDevelopment(type: Copy) {
+  dependsOn tasks.shadowJar
+  from shadowJar.outputs.files.singleFile
+  into file("$buildDir/buildExecutable")
+
+  rename { fileName ->
+    return grpcPluginUberJarName
+  }
+}
+
 task buildExecutable {
-  def outputFile = file("$buildDir/buildExecutable/protoc-gen-servicetalk_grpc.exe")
+  def isWindows = org.gradle.internal.os.OperatingSystem.current().isWindows()
+  def outputFile = new File("$buildDir/buildExecutable/" +
+      addExecutablePostFix("protoc-gen-servicetalk_grpc", isWindows))
+  dependsOn tasks.copyUberJarForDevelopment
   inputs.files shadowJar.outputs.files
   outputs.file outputFile
 
   doLast {
-    if (org.gradle.internal.os.OperatingSystem.current().isWindows()) {
-      throw new GradleException("Windows is currently not supported")
+    if (isWindows) {
+      buildWindowsExecutable(grpcPluginUberJarName, outputFile)
+    } else {
+      buildUnixExecutable(grpcPluginUberJarName, outputFile)
     }
-
-    outputFile.parentFile.mkdirs()
-    outputFile.delete()
-    outputFile <<
-        '''#!/bin/sh
-exec java -jar "$0" "$@"
-exit1
-'''
-    shadowJar.outputs.files.singleFile.withInputStream { outputFile.append(it) }
-    outputFile.executable = true
   }
 }
 tasks.compileJava.finalizedBy(buildExecutable)
 
+task buildExecutableWindowsPublishing {
+  def outputFile = new File("$buildDir/buildExecutable/" +
+      addExecutablePostFix("protoc-gen-servicetalk-windows_grpc", true))
+  dependsOn tasks.copyUberJarForDevelopment
+  inputs.files shadowJar.outputs.files
+  outputs.file outputFile
+
+  doLast {
+    buildWindowsExecutable(grpcPluginUberJarName, outputFile)
+  }
+}
+
+// we attempt to generate both grpc executables when on windows, and don't publish from windows anyways.
+tasks.withType(PublishToMavenRepository) {
+  onlyIf {
+    !org.gradle.internal.os.OperatingSystem.current().isWindows()
+  }
+}
+
 publishing {
   publications {
     mavenJava {
       artifact(buildExecutable.outputs.files.singleFile) {
         classifier = "linux-x86_64"
-        extension = "exe"
+        extension = "sh"
         builtBy buildExecutable
       }
       artifact(buildExecutable.outputs.files.singleFile) {
         classifier = "osx-x86_64"
```

Comment:
What is the expected output for linux and osx classifiers? After running it locally, in `servicetalk-grpc-protoc/build/buildExecutable` folder I see only:
- protoc-gen-servicetalk_grpc.sh
- servicetalk-grpc-protoc-0.26.0-SNAPSHOT-all.jar
Should I see any files with classifiers?

Reply:
Weird, in `servicetalk-grpc-protoc/build/buildExecutable` I see only `protoc-gen-servicetalk_grpc.sh.
But when I do `./gradlew publishToMavenLocal` then I see that it publishes two scripts:
1. `servicetalk-grpc-protoc-0.26.0-SNAPSHOT-linux-x86_64.sh`
2. `servicetalk-grpc-protoc-0.26.0-SNAPSHOT-osx-x86_64.sh`
But their content is identical. I think we may have only one `.sh` script.

Reply:
Tested with the `mavenLocal`, looks like os-classifier is necessary and we need to have separate scripts for all platforms.

Reply:
yes for local development/builds we use a single script targeted at the current platform in `servicetalk-grpc-protoc/build/buildExecutable`. However for maven we publish both scripts with classifiers to disambiguate and this is also what the grpc-gradle-plugin expects too.